x-pack/legacy/plugins/uptime/server/lib/adapters/monitors/elasticsearch_monitors_adapter.ts

```diff
@@ -119,26 +123,45 @@ export class ElasticsearchMonitorsAdapter implements UMMonitorsAdapter {
       statusMaxCount: 0,
     };
 
+    const resultLocations = new Set<string>();
     const linesByLocation: { [key: string]: LocationDurationLine } = {};
-    dateBuckets.forEach(dateBucket => {
-      const x = get(dateBucket, 'key');
-      const docCount = get(dateBucket, 'doc_count', 0);
+    dateHistogramBuckets.forEach(dateHistogramBucket => {
+      const x = get(dateHistogramBucket, 'key');
+      const docCount = get(dateHistogramBucket, 'doc_count', 0);
+      const bucketLocations = new Set<string>();
 
-      dateBucket.location.buckets.forEach(
+      dateHistogramBucket.location.buckets.forEach(
         (locationBucket: { key: string; duration: { avg: number } }) => {
           const locationName = locationBucket.key;
-          let ldl: LocationDurationLine = get(linesByLocation, locationName);
-          if (!ldl) {
-            ldl = { name: locationName, line: [] };
-            linesByLocation[locationName] = ldl;
-            monitorChartsData.locationDurationLines.push(ldl);
+          bucketLocations.add(locationName);
+          resultLocations.add(locationName);
+
+          let currentLine: LocationDurationLine = get(linesByLocation, locationName);
+          if (!currentLine) {
+            currentLine = { name: locationName, line: [] };
+            linesByLocation[locationName] = currentLine;
+            monitorChartsData.locationDurationLines.push(currentLine);
```

Comment:
So this is where you are pushing an empty line for missing duration data?

Reply:
Well, this line was already there but I renamed the variable to `currentLine` because this function was growing and I felt it was a little obscure when trying to read the code.

> So this is where you are pushing an empty line for missing duration data?

The short answer to this question is: "if an empty array is returned as the line data, the UI will render the `EmptyPrompt`.

A longer technical explanation of this code is below:

Basically what's happening here is we can end up with a structure like:
```JSON
"buckets": [
  {
    "@timestamp": 123,
    "locations": [
      {
        "name": "us-east-1",
        "duration": 152
      },
      {
        "name": "us-west-2",
        "duration": 141
      }
    ]
  },
  {
    "@timestamp": 124,
    "locations": [
      {
        "name": "us-east-1",
        "duration": 157
      }
    ]
  },
  {
    "@timestamp": 125,
    "locations": [
      {
        "name": "us-east-1",
        "duration": 150
      },
      {
        "name": "us-west-2",
        "duration": 136
      }
    ]
  },
]
```

In this case, the middle bucket has a missing location. We're [maintaining two `Set`s](https://github.com/elastic/kibana/pull/45734/files#diff-8a8f13ba7685a819cadc890823dd4594R136), one that is reset for each location and another that is used for the entire bucket. This way on each iteration we ensure that [if there's no object for a location that is present elsewhere, we can include an entry in the line with a null](https://github.com/elastic/kibana/pull/45734/files#diff-8a8f13ba7685a819cadc890823dd4594R157) value at the current `x` value (timestamp).

Then, after iterating over a current timestamp's locations, we have less locations than what's in the entire bucket, we "fill in" that location's line (if it exists) with an entry for the timestamp that has a null value.

So while iterating over the data above, on the second timestamp we'd end up with two sets [to evaluate](https://github.com/elastic/kibana/pull/45734/files#diff-8a8f13ba7685a819cadc890823dd4594R149), they'd look like:
```javascript
// location set
['us-east-1']
// histogramBucket set
['us-east-1', 'us-west-2']
```

We then perform a search on the list of lines - if a line exists for the given location, we add an entry. Otherwise we do nothing, as the line won't be rendered anyway for this `x` value.

To be honest, I am not very pleased with the code at the end of this function, it's hard to understand. I will add some comments explaining what's going on, and I'm open to suggestions for cleaning it up.

Reply:
I added comments in https://github.com/elastic/kibana/pull/45734/commits/e9e3e3bb9ab27a386b6b888d91af8b04bdc40f34.